### PR TITLE
Test `metric_attribute` for different children module structures

### LINKED ISSUE #8656: Automatically find metric attribute when logging
### Motivation

When logging metrics stored in a ModuleList, it's somewhat clunky because you have to manually specify the metric_attribute in the call to self.log().

```
def __init__(self, ...):
     self.num_tasks = 50
     self.metrics = nn.ModuleList([torchmetrics.PearsonCorrcoef() for _ in range(num_tasks)])

def test_step(self, data_batch, batch_idx):
     inputs, labels = data_batch
     outputs = self(inputs)

     for task_num in range(self.num_tasks):
          outputs_task = outputs[:, task_num]
          labels_task = labels[:, task_num]
          self.metrics[task_num](outputs_task, labels_task)

          self.log("test/pearson_{}".format(task_num), self.metrics[task_num], metric_attribute="metrics.{}".format(task_num))
```

### Pitch

It seems like Lightning should be able to find the metric_attribute automatically, given the Metric object. This would make logging metrics stored in a ModuleList a bit cleaner. 

### Additional context

---

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #8656

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_metric_result_integration.py [hunk 1]
```diff
@@ -353,3 +354,65 @@ def test_result_collection_extra_reference():
...
+
+            assert results["training_step.e"].meta.metric_attribute == "metrics_collection_dict.a"
+            assert results["training_step.f"].meta.metric_attribute == "metrics_collection_dict.b"
+
+            assert results["training_step.g"].meta.metric_attribute == "metrics_collection_dict_nested.a.0.b"
+            assert results["training_step.h"].meta.metric_attribute == "metrics_collection_dict_nested.a.1"
+
+    model = TestModel()
+
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=2, limit_train_batches=2, limit_val_batches=2)
```
Comment (carmocca) [inline tests/core/test_metric_result_integration.py:417]: Is val important for the test?

```suggestion
    trainer = Trainer(default_root_dir=tmpdir, max_epochs=2, limit_train_batches=2, limit_val_batches=0)
```

**Review by carmocca — APPROVED** [review]

File: tests/core/test_metric_result_integration.py [hunk 2]
```diff
@@ -353,3 +354,65 @@ def test_result_collection_extra_reference():
     """Unit-test to check that the `extra` dict reference is properly set."""
     rc = ResultCollection(True)
     assert rc.extra is rc["_extra"]
+
+
+def test_metric_collections(tmpdir):
+
+    """This test ensures the metric attribute is properly found even with complex nested metric structure"""
```
Comment (awaelchli) [inline tests/core/test_metric_result_integration.py:361]:
```suggestion
def test_metric_collections(tmpdir):
    """This test ensures the metric attribute is properly found even with complex nested metric structure"""
```

**Review by awaelchli — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:
